internal/olm/client/client.go

```diff
@@ -222,7 +224,73 @@ func (c Client) DoCSVWait(ctx context.Context, key types.NamespacedName) error {
 		}
 	}
 
-	return wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
+	err := wait.PollImmediateUntil(time.Second, csvPhaseSucceeded, ctx.Done())
+	if err != nil && errors.Is(err, context.DeadlineExceeded) {
+		if depCheckErr := c.printDeploymentErrors(ctx, key, csv); depCheckErr != nil {
+			return fmt.Errorf("failed to run operator: %v %v", err, depCheckErr)
+		}
+	}
+	return err
+}
+
+// printDeploymentErrors function loops through deployment specs of a given CSV, and prints reason
+// in case of failures, based on deployment condition.
+func (c Client) printDeploymentErrors(ctx context.Context, key types.NamespacedName, csv olmapiv1alpha1.ClusterServiceVersion) error {
+	if key.Namespace == "" {
+		return fmt.Errorf("no namespace provided to get deployment failures")
+	}
+	dep := &appsv1.Deployment{}
+	for _, ds := range csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs {
+		depKey := types.NamespacedName{
+			Namespace: key.Namespace,
+			Name:      ds.Name,
+		}
+		depSelectors := ds.Spec.Selector
+		if err := c.KubeClient.Get(ctx, depKey, dep); err != nil {
+			log.Printf("failed to run operator, deployment not found for : %v\n", ds.Name)
+			continue
+		}
+		for _, s := range dep.Status.Conditions {
```

Comment:
Ideally you shouldn't interleave network calls and print statements; instead all errors should be collected, then printed together. Doing so is ok for now, but can you add a `TODO` to separate these two actions?

Reply:
Sure, so modify these two functions to return errors back to `DoCSVWait()` instead of printing.?

Reply:
Not necessarily. It should look something like (in pseudocode):
```
depErrs := make(map[string]string)
podErrsForDep := make(map[string]map[string]string)
for dep in deployements {
	for cond in dep.Status.Conditions {
		if isStatusNotAvailable(dep) {
			depErrs[dep.Name] = dep.Status.Reason
			podErrsForDep[dep.Name] = getPodErrs(dep)
			break
		}
	}
}

for depName, depReason in depErrs {
	printf("dep %q error: %s\n", depName, depReason)
	for podName, podErr in podErrsForDep[depName] {
		printf("\tpod %q error: %s\n", podName, podErr)
	}
}
```

Reply:
Added TODO comments , and addressed rest of the suggestions.